### ISSUE TITLE
False positive "cannot mutate immutable variable (E0384)" error

### ISSUE THREAD

**Issue by chvllad** [rust-lang/rust-analyzer#16421]
**rust-analyzer version**: 0.4.1816-standalone (d410d4a2b 2024-01-21)

**rustc version**: 1.77.0-nightly (d5fd09972 2024-01-22)

Code https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=143f7fbac71c9d440720ae04f8c9018c gives error `cannot mutate immutable variable f` at the line 21, but compiles fine in rustc.

**Comment by HKalbasi** [comment]
This is strange. If you extract the second closure into a variable, rustc will complain as well. It seems giving a closure directly to the `OneOfFn::B` will unconditionally make it `FnOnce`.

**Comment by strohel** [comment]
I'm also facing similar issue, with (subjectively) simpler reproduction code:
```rust
pub struct Input {
    enabled: bool,
}

pub fn foo(inputs: [&mut Input; 4]) {
    (0..4).for_each(|index| {
        // rust-analyzer complains "cannot mutate immutable variable `inputs`" here
        modify_input(&mut inputs[index].enabled);
    });

    #[allow(clippy::needless_range_loop)]
    for index in 0..4 {
        // No rust-analyzer issue here
        modify_input(&mut inputs[index].enabled);
    }
}

fn modify_input(enabled: &mut bool) {
    *enabled = true;
}
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9d49fd541c1e8f3a64310414aaba3c0d

rust-analyzer v0.4.1830 from https://open-vsx.org/extension/rust-lang/rust-analyzer
